### PR TITLE
Application prefs section

### PR DESCRIPTION
**Pull request by muffinista**
This adds a page to the preferences section to manage credentials for user-created applications.

The goal here is to make it easier for users to manage scripts that require API access. This could be clients, scripts that do some sort of data management, bots, etc. While it's obviously possible to get credentials for a script right now, it's a bit challenging and involved. With this code, you can create an app in the preferences section, and get a token right there. You can also regenerate your token if needed.

These changes also address an issue that exists in the code right now, which is that there is no way to edit the name or website of an app right now.

### CODE REVIEW

File: config/initializers/doorkeeper.rb
Comment (beatrix-bitrot) [inline config/initializers/doorkeeper.rb]: does this have any effects on applications that exist already before these changes are applied?

Comment (muffinista) [inline config/initializers/doorkeeper.rb]: I don't believe so, and I'm running it in production with no reported issues.

Comment (beatrix-bitrot) [inline config/initializers/doorkeeper.rb]: that was my only concern really. there are some merge conflicts so i'm going to mark this as requesting changes

**Review by beatrix-bitrot — CHANGES_REQUESTED** [review]

**Comment by muffinista** [comment]
OK, conflicts should be resolved, thanks!

**Review by beatrix-bitrot — APPROVED** [review]

**Review by beatrix-bitrot — DISMISSED** [review]
the CI checks are failing ):

**Comment by muffinista** [comment]
So sorry about that! I'm travelling and made a typo using the github inline editor. CI is passing now.

**Comment by beatrix-bitrot** [comment]
No worries. Also sorry to ask but could you add a screenshot or two? I'm trying to take a little more care with my reviews and so I'd either want to run your changes on one of my servers before approving or at least see how it looks in some screenshots.

**Comment by muffinista** [comment]
No problem at all. First, the code is running on [botsin.space](https://botsin.space) if you want to create an account.


Here's the app list:

screenshot 2017-05-10 20 41 16

Here's the Create New App page:

screenshot 2017-05-10 20 42 02

The app page, with tokens:
screenshot 2017-05-10 20 42 57

**Review by akihikodaki — CHANGES_REQUESTED** [review]
It may be good to have a notice to tell users to keep tokens _secret_ to the application page.
Mastodon has a variety of users and some of them are not familiar with them.

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: Spec that it acutally shows the application. (testing `assigns(:application)` is enough.)

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: * Test if it actually creates an application.
* Use `context` instead of `describe`.

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: * Test if it actually updates an application.
* Use `context` instead of `describe`.

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: It is not `PUT` but `PATCH`.

**Comment by nightpool** [comment]
Is there a reason the redirect URI is a textarea instead of just an input? the styling looks odd.

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: Do not add spaces in parentheses.

File: spec/models/user_spec.rb
Comment (akihikodaki) [inline spec/models/user_spec.rb]: Use `update!` instead of a combination of an assignment and `save!`.

File: config/initializers/doorkeeper.rb
Comment (akihikodaki) [inline config/initializers/doorkeeper.rb]: I'm not sure even why `owner` is required. Could you explain the reason?

**Review by akihikodaki — CHANGES_REQUESTED** [review]
For reuqests of additional tests; [pending examples](https://relishapp.com/rspec/rspec-core/v/3-6/docs/pending-and-skipped-examples) would be sufficient to merge this change, but actual examples will be appreciated.

Thank you for your contribution. This feature is essential.

File: config/initializers/doorkeeper.rb
Comment (muffinista) [inline config/initializers/doorkeeper.rb]: Yes, it sets the application owner to the current user if that data is available. Without that, we can't associate apps with owners.

Comment (akihikodaki) [inline config/initializers/doorkeeper.rb]: I mean I doubt associating apps with owners is necessary. Isn't tracking access tokens with resource owners enough?

Comment (muffinista) [inline config/initializers/doorkeeper.rb]: Maybe I'm missing something. The whole point of this PR is to associate applications with their creator so they can be managed and displayed in the settings section of that user. While tokens have an associated user, its very difficult to use that link to actually do any management of the app without some brittle hacks.

Comment (akihikodaki) [inline config/initializers/doorkeeper.rb]: I see. It looks good to me to add `owner`.

**Comment by muffinista** [comment]
Thanks for the feedback! I think I've addressed most of the concerns raised here. I've updated the specs, added a note about not sharing API keys, and made a few other changes. I didn't change the textarea for the redirect URL for while I agree with @nightpool that the textarea looks a little weird here (especially with the default redirect), it makes it a lot easier to see URLs that are longer. I would be fine with changing it to a normal input though.

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: Use `update!` here, too.

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: I think `let!` is what you expect, isn't it?

File: spec/controllers/settings/applications_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/settings/applications_controller_spec.rb]: This is trivial, but `be nil` is used here while the other part uses `be_nil`. Anyway we do not have standard about that as far as I know, so it does not blocking merging.

**Review by akihikodaki — CHANGES_REQUESTED** [review]
I request some minor changes. This change will be ready to merge after they are addressed.

**Review by akihikodaki — APPROVED** [review]

**Comment by muffinista** [comment]
@beatrix-bitrot @akihikodaki pinging to check if/when this might be merged, and if anything else is needed to help with that? thanks!

**Comment by beatrix-bitrot** [comment]
@Gargron please merge, request changes, or close this~

File: db/migrate/20170427011934_re_add_owner_to_application.rb
Comment (Gargron) [inline db/migrate/20170427011934_re_add_owner_to_application.rb]: Wait, does this have to be polymorphic? We're never gonna attach the application to anything other than `User`. If this could be only owner_id with a foreign key ensuring integrity with the users table, that'd be far better.

**Review by Gargron — CHANGES_REQUESTED** [review]

File: db/migrate/20170427011934_re_add_owner_to_application.rb
Comment (muffinista) [inline db/migrate/20170427011934_re_add_owner_to_application.rb]: Adding the foreign key is no problem.

I'm reluctant to remove the polymorphic association because it's part of `Doorkeeper` and the underlying code expects it, and removing it could have some weird side-effects. At the very least, removing `owner_type` breaks a bunch of specs I wrote which rely on [the Application fabricator](https://github.com/muffinista/mastodon/blob/application-prefs-section/spec/fabricators/application_fabricator.rb) and while I can probably figure out a way around that issue, it seems like it would introduce a lot of potential fragility.

Comment (Gargron) [inline db/migrate/20170427011934_re_add_owner_to_application.rb]: Alright, that will do for now, since owner_type is essentially redundant.

**Review by Gargron — APPROVED** [review]

File: db/migrate/20170427011934_re_add_owner_to_application.rb
Comment (muffinista) [inline db/migrate/20170427011934_re_add_owner_to_application.rb]: Thank you!

**Pull request merged**
